x-pack/plugins/snapshot_restore/public/application/components/policy_form/steps/step_settings/fields/indices_and_data_streams_field/indices_and_data_streams_field.tsx

```diff
@@ -0,0 +1,347 @@
+      Array.isArray(config.indices) && !isAllIndices
+        ? indicesAndDataStreams.filter((i) => (config.indices! as string[]).includes(i))
+        : [...indicesAndDataStreams]
+  );
+
+  // States for choosing all indices, or a subset, including caching previously chosen subset list
+  const [indicesAndDataStreamsOptions, setIndicesAndDataStreamsOptions] = useState<
+    EuiSelectableOption[]
+  >(() =>
+    mapSelectionToIndicesOptions({
+      selection: indicesAndDataStreamsSelection,
+      dataStreams,
+      indices,
+      allSelected: isAllIndices || typeof config.indices === 'string',
+    })
+  );
+
+  // State for using selectable indices list or custom patterns
+  const [selectIndicesMode, setSelectIndicesMode] = useState<'list' | 'custom'>(() =>
+    determineListMode({ configuredIndices: config.indices, dataStreams, indices })
+  );
+
+  // State for custom patterns
+  const [indexPatterns, setIndexPatterns] = useState<string[]>(() =>
+    typeof config.indices === 'string'
+      ? (config.indices as string).split(',')
+      : Array.isArray(config.indices) && config.indices
+      ? config.indices
+      : []
+  );
+
+  const indicesSwitch = (
+    <EuiSwitch
+      label={
+        <FormattedMessage
+          id="xpack.snapshotRestore.policyForm.stepSettings.allDataStreamsAndIndicesLabel"
+          defaultMessage="All data streams and indices, including system indices"
+        />
+      }
+      checked={isAllIndices}
+      disabled={isManagedPolicy}
+      data-test-subj="allIndicesToggle"
+      onChange={(e) => {
+        const isChecked = e.target.checked;
+        setIsAllIndices(isChecked);
+        if (isChecked) {
+          setIndicesAndDataStreamsSelection(indicesAndDataStreams);
+          setIndicesAndDataStreamsOptions(
+            mapSelectionToIndicesOptions({
+              allSelected: isAllIndices || typeof config.indices === 'string',
+              dataStreams,
+              indices,
+              selection: indicesAndDataStreamsSelection,
+            })
+          );
+          onUpdate({ indices: undefined });
+        } else {
+          onUpdate({
+            indices:
+              selectIndicesMode === 'custom'
+                ? indexPatterns.join(',')
+                : [...(indicesAndDataStreamsSelection || [])],
+          });
+        }
+      }}
+    />
+  );
+
+  return (
+    <EuiDescribedFormGroup
+      title={
+        <EuiTitle size="s">
+          <h3>
+            <FormattedMessage
+              id="xpack.snapshotRestore.policyForm.stepSettings.dataStreamsAndIndicesTitle"
+              defaultMessage="Data streams and indices"
+            />
+          </h3>
+        </EuiTitle>
+      }
+      description={
+        <FormattedMessage
+          id="xpack.snapshotRestore.policyForm.stepSettings.dataStreamsAndIndicesDescription"
+          defaultMessage="To back up indices and data streams, manually select them or define index patterns to dynamically capture them."
```

Comment:
Nit: I think `match` is a better term here than `capture`.
```suggestion
          defaultMessage="To back up indices and data streams, manually select them or define index patterns to dynamically match them."
```